src/RadixDlt.NetworkGateway.PostgresIntegration/Services/Queries/MetadataPageQuery.cs

```diff
@@ -0,0 +1,255 @@
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using RadixDlt.NetworkGateway.Abstractions.Extensions;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models.CustomTypes;
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using GatewayModel = RadixDlt.NetworkGateway.GatewayApiSdk.Model;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.Services.Queries;
+
+internal static class MetadataPageQuery
+{
+    internal readonly record struct QueryConfiguration(
+        GatewayModel.IdBoundaryCoursor? Cursor,
+        int MaxPageSize,
+        int MaxDefinitionsToRead
+    );
+
+    private readonly record struct QueryResultRow(
+        long EntityId,
+        long TotalEntriesExcludingDeleted,
+        long TotalEntriesIncludingDeleted,
+        long DefinitionId,
+        string Key,
+        long KeyFirstSeenStateVersion,
+        byte[] Value,
+        bool IsLocked,
+        bool IsDeleted,
+        long LastUpdatedStateVersion,
+        bool FilterOut,
+        bool IsLastCandidate,
+        IdBoundaryCursor? NextCursorExclusive
+    );
+
+    internal static async Task<Dictionary<long, GatewayModel.EntityMetadataCollection>> ReadPages(
+        DbConnection dbConnection,
+        IDapperWrapper dapperWrapper,
+        GatewayApiSdk.Model.LedgerState ledgerState,
+        long[] entityIds,
+        QueryConfiguration queryConfiguration,
+        byte networkId,
+        CancellationToken token = default)
+    {
+        var queryParameters = new
+        {
+            entityIds = entityIds,
+            useCursor = queryConfiguration.Cursor is not null,
+            atLedgerState = ledgerState.StateVersion,
+            cursorStateVersion = queryConfiguration.Cursor?.StateVersionBoundary ?? 0,
+            cursorDefinitionId = queryConfiguration.Cursor?.IdBoundary ?? 0,
+            pageLimit = queryConfiguration.MaxPageSize,
+            definitionReadLimit = Math.Floor(queryConfiguration.MaxDefinitionsToRead / (decimal)entityIds.Length),
+        };
+
+        const string QueryText = @"
+WITH vars AS (
+    SELECT
+        unnest(@entityIds) AS entity_id,
+        CAST(@useCursor AS bool) AS use_cursor,
+        ROW(CAST(@cursorStateVersion AS bigint), CAST(@cursorDefinitionId  AS bigint)) AS start_cursor_exclusive,
```

Comment:
I wonder if we can just provide an IdBoundaryCursor here, and cast it as Row(bigint, bigint) or something? :)

Reply:
unfortunately it'll not be that easy (there is no support for custom types in entity framework), spent some time trying to do that in a hacky way and since it has little benefit i'm gonna stop wasting time.